### DISCUSSION TITLE
Custom types not passed to constructor on rehydration

### DISCUSSION THREAD

**Discussion by mjroeleveld** [mikro-orm/mikro-orm#4672] [General]
It feels rather unexpected that with an embeddable, properties are passed as primitives to the constructor, to be only converted to their actual expected type in a later stage.

```
@Embeddable()
export class Test {
  @Property({ type: DateType })
  public dateProp: Date;

  constructor({
    dateProp
  }: {
    dateProp: Date;
  }) {
    //  This throws, because dateProp is a string at this stage.
    if (dateProp.getTime() > Date.now()) {
      throw new Error('invalid date');
    }

    this.dateProp = dateProp;
  }
}
```

I would expect `dateProp` to be a `Date` in the constructor. Note that I am using `forceEntityConstructor: true`.

**Comment by B4nan** [comment]
>   @Property({ type: DateType })

`DateType` is for mapping dates (no time info), not datetimes, do you understand that? I think I will rework it in v6 so it just maps to a string instead of `Date`, it was never meant to be something for production usage, rather an example type for others to see how it works, it makes very little sense to map a date to the `Date` object imo.

What version are you on? What driver? How does the entity definition look like? Is this inlined or object embeddable? Ideally provide complete repro.

**Reply by mjroeleveld** [reply]
Good point, yes, I was aware of that. I am on `5.7.14` with a Postgres driver. It is not inlined, but used in an embedded property. I realise `Date` is not the best example, but I am using it because it's closest to the documentation. I would expect the mapped type in the constructor though.

**Reply by B4nan** [reply]
Yeah, it should be converted. But with `Date` objects it's much more complicated, as it is not the ORM or the type that converts them, it's the database layer (`node-postgres` in this case). And for object embeddable, it does not work this way. 

> Note that I am using forceEntityConstructor: true.

Curious to hear why? So far I have seen quite a lot of people who used it without understanding what it does (and in fact they did not needed it). IMO this toggle makes sense only if you want to use native private properties in your entity. Note that new entities are always created via the constructor, this flag is only about creating managed entity instances.

**Reply by B4nan** [reply]
FYI I am already in the middle of a refactoring for v6 that will rework the `Date` objects mapping so it is handled by the ORM type classes, you can subscribe to #4603. In fact, this is partially already reworked in v6, so maybe it already works in there - v6 no longer uses the database layer for the conversion and handles it explicitly (but not via the type classes).

**Reply by mjroeleveld** [reply]
Okay great!

> Curious to hear why?

Because I have my domain logic on the ORM objects, making use of methods and private properties to do all kind of stuff (e.g. `entity.increaseCounter()`). Unfortunately this is not working so great because hydration bypasses the constructor anyway for non-primitive types. All properties need to be public because of that and encapsulation is only theoretical (any property can be reassigned directly: `entity.counter++`). I understand this is because of Typescript's limitations. Apart from this Mikro ORM has been great for this use case 😄.

**Reply by B4nan** [reply]
Can you provide a complete example? Maybe entity definition should be enough so I can better understand. By private properties you mean real JS private props (prefixed with `#`) or TS private props (so just `private` modifier)?

**Reply by mjroeleveld** [reply]
The idea is that you protect invariants by having the domain logic on your data objects. E.g. meaning that you cannot decrement a counter to a negative number. This way you ensure system data integrity because no one can alter the data directly (by editing properties) except for calling methods on the domain model. The counterexample (and antipattern) of this is the [anemic domain model](https://en.wikipedia.org/wiki/Anemic_domain_model?useskin=vector), where domain logic often lives in controllers etc.

A real-life example of this in my case:
```typescript
@Entity()
export class Car implements IAggregateRoot {
  @PrimaryKey()
  public readonly id: string;

  @OneToMany({
    entity: () => DriverGroup,
    mappedBy: 'car',
  })
  public readonly driverGroups: Collection<DriverGroup> = new Collection<DriverGroup>(
    this,
  );

  addDriverGroup({ id, name }: {}) {
    // Only add the group if there is no more than 3, make sure the names are unique, etc.
  }

  constructor(...) {
    // ...
  }
}
```

The `addDriverGroup` protecting the invariants on the `Car` entity is basically useless because I can just go ahead and do `new Car().driverGroups.add(...)` bypassing the invariants (`driverGroups` is not private). The same counts for simple properties like dates, strings and anything else that needs setters.

**Reply by B4nan** [reply]
Ok, I can understand that, but I don't see why would you need the `forceEntityConstructor` for this? I mean it will be exactly the same problem regardless of it, right? I wanted to see an example of where you need this flag to be enabled.

**Reply by mjroeleveld** [reply]
Yes, but using logic-heavy data objects means that you need private properties down the road, and then you need `forceEntityConstructor`, right?

**Reply by B4nan** [reply]
Once again, the `forceEntityConstructor` is required only if you want to use native private properties (e.g. `#foo`), it is irrelevant for the TS `private` modifier (e.g. `private foo`).

**Reply by mjroeleveld** [reply]
Ahh, right 😅. It's not essential to the initial discussion though. You can forget about my note on this flag.

**Reply by mjroeleveld** [reply]
FYI: I just tried to disable the flag, but instance methods are not initialized on hydrated objects without the flag. `addDriverGroup` in the example above.

**Reply by B4nan** [reply]
What do you mean by that? Initialized methods?

**Reply by mjroeleveld** [reply]
Methods/properties defined on the entity class other than the ORM decorated properties are not defined after hydration.

**Reply by B4nan** [reply]
Well, you get the very same class instance regardless of this flag, with all the methods. If you don't, provide a complete repro, but I can assure you, this is not what the `forceEntityConstructor` is about.

Also, there is no such thing as "initialized method", all class methods are defined on the class prototype, and you get entities created with this prototype (again, regardless of the `forceEntityConstructor` flag).

**Reply by B4nan** [reply]
FYI the flag controls how the instances are created for managed entities (those loaded from database), by default managed entities are created via `Object.create()` which won't trigger the constructor, that is literally the only difference:

```ts
class Foo {
    test() {
        console.log('bar');
    }
}

const f1 = new Foo();
f1.test();

const f2 = Object.create(Foo.prototype, {});
f2.test();
```

**Reply by mjroeleveld** [reply]
Sorry @B4nan it took me a while to get back to it. I needed to do more digging. It seems that embeddables do **not** receive data through the constructor like entities but that properties are instead assigned after the constructor is called? Is this correct?

**Reply by B4nan** [reply]
It works the same way as with entities, constructors are fired only for new embeddables (e.g. when you create them with `em.create()` or when you fire the constructor yourself via `new`), not for those on managed entities (so loaded from the database). The point behind this is that you can use property initializers for things like `createdAt` dates - as they will be fired only when you actually create the entity/embeddable.

**Reply by mjroeleveld** [reply]
Alright, just to confirm, is this the same when `forceEntityConstructor` is `true`?

**Reply by B4nan** [reply]
Nope, I already told you this flag is exactly about controlling this behavior, when set to true, the ORM will use constructors for the managed entities too, that's why it's called "force entity constructors", as it forces the ORM to use them all the time, not just for new entities ("new" as in "those that will trigger insert queries").

**Reply by mjroeleveld** [reply]
Yes, I though so, but I wanted to confirm because I didn't know after your last comment. I am trying to make a repro for what I'm running into but I can't seem to reproduce it in a test. Sometimes I see with my debugger that properties have unexpected values like `null` which indicates the constructor is not called even though `useEntityConstructor` is `true`. It doesn't make sense and it's why I am so confused.
